feat(router): add routes to RouteConfigLoadEnd (#16282)

meeroslav opened the pull request:
**Please check if the PR fulfills these requirements**

* [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
* [X] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?**
When accessing modules lazily, RouteConfigLoadEnd gets fired but before routes are loaded and assigned to route._loadedConfig and therefore accessible.
More details in: [https://github.com/angular/angular/issues/16282](https://github.com/angular/angular/issues/16282)

**What is the new behavior?**
RouteConfigLoadEnd is fired once routes are fetched and has routes as a second parameter.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

Commit:
feat(core): allow custom selector when bootstrapping components (#15668)

- fixes #7136

PR Close #15668

Commit:
feat(forms): introduce min and max validators (#15813)

PR Close #15813

ocombe committed:
feat(compiler-cli): add param to set MissingTranslationStrategy on ngc (#15987)

This commit adds a new parameter to ngc named `missingTranslation`  to set the MissingTranslationStrategy for AoT, it takes the value `error`, `warning` or `ignore`.

Fixes #15808

PR Close #15987

sjtrimble committed:
fix(aio): resource nav

gkalpak committed:
fix(upgrade): use correct attribute name for upgraded component's bindings (#16128)

Previously, when using a different property/attribute name for an upgraded
component's binding (e.g. `bindings: {propName: '<attrName'}`), the property and
attribute names were swapped (e.g. using `attrName` as the property name and
`propName` as the attribute name). This resulted in unexpected behavior.

This commit fixes this using the correct names for properties and attributes.

This only affects `upgrade/dynamic`. `upgrade/static` works correctly already.

Fixes #8856

PR Close #16128

vicb committed:
fix(router): forward the query parameters in the ng1 -> ng2 url sync (#16249)

fixes #16067

PR Close #16249

petebacondarwin committed:
build(aio): fix rendering of Decorator API docs

The templates were a bit out and we were not merging the
correct properties.

Added some docs and tests for the processor.

Related to #16208

petebacondarwin committed:
build(aio): include missing packages in the docs

These packages were not included in the configuration of the doc-gen, so
were not being rendered.

petebacondarwin committed:
build(aio): fix other dgeni templates

The function and interface templates were overriding the
wrong block.

swseverance committed:
docs(aio): remove duplicate `as` (#16265)

the word `as` incorrectly appears consecutively

cmckni3 committed:
docs(aio): Fix decorator JS feature explanation (#16214)

sjtrimble committed:
feat(aio): api page column layout

Foxandxss committed:
docs(aio): remove _blank target (#16205)

sjtrimble committed:
feat(aio): copy code snackbar and design updates

- Add snackbar and pointer cursor for copy code button inside aio-code components
- Flex cenetered content in features page
- Removed duplicate global css class
- Add styles to links inside of sub-sections
- Remove focus outline on top nav bar links

petebacondarwin committed:
build(aio): ignore the VERSION constant correctly

We were filtering this document from the docs list but
not removing it from the module export lists.

We can actually filter it out much easier at the TypeScript
parsing point, which means we do not need the
`filterIgnoredDocs` processor any more.

Closes #16287

wardbell committed:
fix(aio): AppComponent should scroll only once when location changes

wardbell committed:
feat(aio): don't animate sidenav on launch.

sjtrimble committed:
fix(aio): copy button placement fix

Added material-icons back in used throughout codebase prevoiusly renamed

sjtrimble committed:
feat(aio): layout max width and design cleanup

- Content container now has a max width and made the body background the same offwhite color so the cut-off is not visible
- Sidenav will always remain at the left of the page
- Added a max width to the api banner of filter search inputs
- Left aligned content container so it is flush with sidenav
- Changed API filter width to match size of filter results

meeroslav committed:
feat(router): add routes to RouteConfigLoadEnd (#16282)

meeroslav committed:
feat(router): add routes to RouteConfigLoadEnd

tests

meeroslav commented:
I signed it!

meeroslav committed:
feat(router): add routes to RouteConfigLoadEnd

fix tslint

petebacondarwin committed:
test(aio): add missing test for AppComponent 'starting' feature

meeroslav committed:
feat(router): add routes to RouteConfigLoadEnd

formatting fix

meeroslav committed:
feat(router): add routes to RouteConfigLoadEnd - fix test

petebacondarwin committed:
build(aio): turn on dangling link checking

gkalpak committed:
refactor(aio): move unused images to `unused/` sub-directories

gkalpak committed:
fix(aio): fix ServiceWorker routing

gkalpak committed:
fix(aio): update files ignored by the ServiceWorker

Fixes #16198

gkalpak committed:
feat(aio): cache external resources in the ServiceWorker

Cache resources that are necessary for displaying the basic pages when offline.

wardbell committed:
feat(aio): boilerplate:add cleans (removes) before adding

sjtrimble committed:
feat(aio): api pages styling

- Banner class code consolidation for API pages
- Set up temporary table of contents class and file
- API pages title styling
- Add color styling to doc-type listed on each API details page
- Classes and SCSS variables refactor
- Mobile optimization on headers, info-banner, and API pages
- API page table custom styling
- API Class Overview template code overview into table format

chuckjaz committed:
feat(language-service): provide external file list to TypeScript (#16417)

Also ensures that it only calls base language service for files
that it contains.

PR Close #16417

gkalpak committed:
build(aio): update @angular/service-worker to v1.0.0-beta.10

wardbell committed:
feat(aio): revise Docs page; docs version selector in sidenav

vicb commented on packages/router/src/router_config_loader.ts:
```diff
@@ -35,13 +35,14 @@ export class RouterConfigLoader {
     const moduleFactory$ = this.loadModuleFactory(route.loadChildren !);
 
     return map.call(moduleFactory$, (factory: NgModuleFactory<any>) => {
+      const module = factory.create(parentInjector);
```
What we want to measure the the load time and here you incidentally add the module creation time.

Adding a new event type would be better IMO.

meeroslav replied:
I didn't know that part. It makes sense.

Should I propose a solution with new event type e.g. RouteChildrenLoaded and revert current changes or does this have to be further discussed within the team first?

gkalpak committed:
refactor(aio): move content-related images to `content/images/`

gkalpak committed:
fix(aio): fix incorrect image URLs in guides

gkalpak committed:
refactor(aio): move `/guide/docs.md` to `/marketing/docs.md`

Because:
1. `docs` feels like a "top level" page, similar to `features`, `resources`, `events` etc.
2. This enables the ServiceWorker to pre-fetch/cache the document (similar to what happens with all
   other direct children of `content/docs/`), without the need for special-casing it in
   `ngsw-manifest.json`.

mhevery committed:
release: cut the 4.1.0 release

mhevery committed:
docs: add changelog for 4.1.0

petebacondarwin committed:
build(aio): move attribute utils to helpers folder

This allows these utility functions to be reused across packages.

petebacondarwin committed:
build(aio): include new packages in docs-watch

mhevery committed:
docs: update 4.1.0 changelog, remove aio

mhevery committed:
docs: update 4.1.0 changelog, remove aio

petebacondarwin committed:
fix(aio): ensure styling of icons is specific to usage

Previously the CSS styling for material icons was too broad and affected
all instances of icons. This commit constrains the position of copy button
icons only to copy buttons.

petebacondarwin committed:
fix(aio): improve positioning of heading-anchors

petebacondarwin committed:
build(aio): update dgeni-packages to 0.17.2 (#16358)

This version has fix for the jsdoc parser, which was
causing false negative warnings and errors in the doc
gen.

Closes #16319

petebacondarwin committed:
build(aio): add new post-process dgeni package

This package will allow us to do complex post-processing
on the HTML that is rendered by Nunjucks.

petebacondarwin committed:
build(aio): move autolink-headings to post-processing (#16336)

The autolinking is now done on the `renderedContent` which means it also
captures and autolinks headings that were generated outside of markdown.

PR Close #16336

mhevery committed:
fix: strictNullCheck support. (#16389) (#16389)

Fix #16357

Workaround for https://github.com/Microsoft/TypeScript/issues/10078

Closes #16389

PR Close #16389

wardbell committed:
feat(aio): add Table of Contents (toc) component. (#16078)

wardbell committed:
feat(aio): implement gkalpak review suggestions

tbosch committed:
fix(core): allow directives to inject the component’s `ChangeDetectorRef`. (#16394)

When a directive lives on the same element as a component
(e.g. `<my-comp myDir>`), the directive was not able to get hold
of the `ChangeDetectorRef` of the component on that element. However,
as directives are supposed to decorate components, this is incorrect.

This commit enables this use case.

Closes #12816

tbosch committed:
fix(core): allow to detach `OnPush` components (#16394)

Fixes #9720

tbosch committed:
fix(core): don’t set `ng-version` for dynamically created components (#16394)

Angular uses the `ng-version` attribute to indicate which elements
were used to bootstrap an application. However, after 4.0 we also
added this attribute for all dynamically created components.

Fixes #15880

PR Close #16394

swseverance committed:
docs(aio): correct abbreviation (#16295)

change abbreviation from 'ES' to 'ES2015'. the 'ES2015' abbreviation is used elsewhere in the tutorial.

wardbell committed:
feat(aio): code snippet source available & shown when code missing

Tells reader (usually the author) what code file is missing
Also when no linenums specified, turn them on if num of lines > 10

petebacondarwin committed:
build(aio): move copyContentAssets processor to the base package

This allows other processors who need to know about the copyContentAssets
processors to ensure that the runs after the content has been copied.

marclaval committed:
feat(core): upgrade dep on zone.js to 0.8.9 (#16401)

PR Close #16401

petebacondarwin committed:
build(aio): capture and log errors and warnings when post-processing HTML

gkalpak committed:
fix(aio): do not create ToC for `file-not-found` and `fetching-error` documents

Related to #16078.

petebacondarwin committed:
style(aio): fix comment on postProcessHtml processor

Fixes https://github.com/angular/angular/pull/16336#discussion_r113891332

vicb committed:
docs(common): fix API docs for NgComponentOutlet (#16411)

fixes #16373

PR Close #16411

petebacondarwin committed:
build(aio): auto-fill width/height to all image tags

Parse all `<img>` tags, during doc-gen, and insert the width and height of
the sourceed image, if neither are already specified.

Warnings are reported if the `<img>` tag has no `src` attribute or the image
cannot be loaded.

The work is done in the `addImageDimensions` post-processor, which must be
configured with a `basePath` so that it knows where to find the images.

Closes #15888

petebacondarwin committed:
docs(aio): fix image size on homepage

vicb committed:
fix: public API golden files (#16414)

chuckjaz committed:
ci(language-service): update ci tests to official 2.3 build (#16415)

PR Close #16415

tbosch committed:
fix(core): don’t stop change detection because of errors

- prevents unsubscribing from the zone on error
- prevents unsubscribing from directive `EventEmitter`s on error
- prevents detaching views in dev mode if there on error
- ensures that `ngOnInit` is only called 1x (also in prod mode)

Fixes #9531
Fixes #2413
Fixes #15925

chuckjaz committed:
fix(language-service): remove asserts for non-null expressions (#16422)

Reworked some of the code so asserts are no longer necessary.
Added additional and potentially redundant checks
Added checks where the null checker found real problems.

PR Close #16422

kimburgess committed:
docs(aio): fix broken NgModules link in tutorial (#16398)

Brackets included in link title contents were breaking markdown
parsing, preventing correct link generation.

willshowell committed:
docs(aio): add missing word (#16328)

kapunahelewong committed:
docs(aio): apply guidelines to ts-to-js - adapt for aio site (#16054)

wardbell committed:
feat(aio): add marketing class to AppComponent <aio-shell> on mkt page (#16395)

Also navigation.json Doc page should be hidden in sidebar, shown in top

wardbell committed:
feat(aio): add tooltip to nav node if none specified

petebacondarwin committed:
build(aio): remove unused `{@exampleTabs}` inline tag

petebacondarwin committed:
build(aio): convert triple backtick blocks to `code-example` tags

Closes #16246

wardbell committed:
feat(aio): tell user when no search results

closes #16294

wardbell committed:
docs(aio): Add numbers and steps to Tutorial navigation (#16421)

wardbell committed:
fix(aio): header click should always toggle expand/collapse (#16439)

Also adds class tests for this NavItemComponent.
Not (yet) testing effects on the templated HTML

wardbell committed:
test(aio): confirm ga tracks user’s # fragment clicks within a page (#16441)

This was the behavior and we think we still want to do it.
Added comment and test to confirm that is our present intention.

Foxandxss committed:
docs(aio): fix lots of broken links

Closes #16317

petebacondarwin committed:
build(aio): improve info-bar template

The "npm module" and "NgModule" are now rendered correctly.

Closes #16055

sjtrimble committed:
fix(aio): correct position of sidenav icons

sjtrimble committed:
fix(aio): remove max width on marketing pages and footer

- Removed the max-width on all marketing pages
- Moved the footer so that it extends fully horizontally on all screen sizes

sjtrimble committed:
fix(aio): search result styling

- Changed search result link hover state to white to be more legible
- Increased the max-height on search results container to remove the jitter
- Changed search results to have space-around vs space-between for more appropriate spacing

jasonaden commented:
@meeroslav Thanks for this PR and sorry you didn't get an answer earlier.

If we want to add this feature, I think you should add a new event. But can you first create an issue as a feature request so this could be discussed? Put in there what the feature is, but also the use case. It's easier to determine the best way to handle a problem if we know what the problem is that you're trying to solve.

I'll close this PR so you can open the issue and we can iterate there first.

meeroslav commented:
@jasonaden The issue was already opened 3 months ago and mentioned [here](https://github.com/angular/angular/issues/16282) and [here](https://github.com/angular/angular/issues/14036). Also, package that depends on this functionality is `localize-router` with appropriate issue mentioned [here](https://github.com/Greentube/localize-router/issues/45).

Please let me know if you need more information.

If it's easier I can create new pull request based on @vicb recommendation.
